doc/sr_post.1.rst

=========
 SR_Post
=========

-------------------------------------------------
Publish the Availability of a File to Subscribers
-------------------------------------------------

:Manual section: 1 
:Date: @Date@
:Version: @Version@
:Manual group: MetPx Sarracenia Suite

.. contents::

SYNOPSIS
========

**sr_post|sr_cpost** [ *OPTIONS* ][ *-pb|--post_broker broker* ][ *-pbu|--post_base_url url[,url]...** ] 
[ *-p|--path ] path1 path2...pathN* ]

( also **libsrshim.so** )

DESCRIPTION
===========

**sr_post** posts the availability of a file by creating an announcement.
In contrast to most other sarracenia components that act as daemons,
sr_post is a one shot invocation which posts and exits.
Subscribers use `sr_subscribe <sr_subscribe.1.rst>`_  

To make files available to subscribers, **sr_post** sends the announcements 
to an AMQP server, also called a broker.  

This manual page is primarily concerned with the python implementation,
but there is also an implementation in C, which works nearly identically.
Differences:

 - plugins are not supported in the C implementation.
 - C implementation uses POSIX regular expressions, python3 grammar is slightly different.
 - when the *sleep* option ( used only in the C implementation) is set to > 0,
   it transforms sr_cpost into a daemon that works like `sr_watch(1) <sr_watch.1.rst>`_.  


Mandatory Settings
------------------

The [*-pbu|--post_base_url url,url,...*] option specifies the location 
subscribers will download the file from.  There is usually one post per file.
Format of argument to the *post_base_url* option::

       [ftp|http|sftp]://[user[:password]@]host[:port]/
       or
       file:

When several urls are given as a comma separated list to *post_base_url*, the
url´s provided are used round-robin style, to provide a coarse form of load balancing.

The [*-p|--path path1 path2 .. pathN*] option specifies the path of the files
to be announced. There is usually one post per file.
Format of argument to the *path* option::

       /absolute_path_to_the/filename
       or
       relative_path_to_the/filename

The *-pipe* option can be specified to have sr_post read path names from standard 
input as well.


An example invocation of *sr_post*::

 sr_post -pb amqp://broker.com -pbu sftp://stanley@mysftpserver.com/ -p /data/shared/products/foo 

By default, sr_post reads the file /data/shared/products/foo and calculates its checksum.
It then builds a post message, logs into broker.com as user 'guest' (default credentials)
and sends the post  to defaults vhost '/' and default exchange. The default exchange 
is the prefix *xs_* followed by the broker username, hence defaulting to 'xs_guest'.
A subscriber can download the file /data/shared/products/foo by authenticating as user stanley
on mysftpserver.com using the sftp protocol to broker.com assuming he has proper credentials.
The output of the command is as follows ::

 [INFO] Published xs_guest v02.post.data.shared.products.foo '20150813161959.854 sftp://stanley@mysftpserver.com/ /data/shared/products/foo' sum=d,82edc8eb735fd99598a1fe04541f558d parts=1,4574,1,0,0

In MetPX-Sarracenia, each post is published under a certain topic.
The log line starts with '[INFO]', followed by the **topic** of the
post. Topics in *AMQP* are fields separated by dot. The complete topic starts with
a topic_prefix (see option), version *V02*, an action *post*,
followed by a subtopic (see option) here the default, the file path separated with dots
*data.shared.products.foo*.

The second field in the log line is the message notice.  It consists of a time 
stamp *20150813161959.854*, and the source URL of the file in the last 2 fields.

The rest of the information is stored in AMQP message headers, consisting of key=value pairs.
The *sum=d,82edc8eb735fd99598a1fe04541f558d* header gives file fingerprint (or checksum
) information.  Here, *d* means md5 checksum performed on the data, and *82edc8eb735fd99598a1fe04541f558d*
is the checksum value. The *parts=1,4574,1,0,0* state that the file is available in 1 part of 4574 bytes
(the filesize.)  The remaining *1,0,0* is not used for transfers of files with only one part.

Another example::

 sr_post -pb amqp://broker.com -pbd /data/web/public_data -pbu http://dd.weather.gc.ca/ -p bulletins/alphanumeric/SACN32_CWAO_123456

By default, sr_post reads the file /data/web/public_data/bulletins/alphanumeric/SACN32_CWAO_123456
(concatenating the post_base_dir and relative path of the source url to obtain the local file path)
and calculates its checksum. It then builds a post message, logs into broker.com as user 'guest'
(default credentials) and sends the post to defaults vhost '/' and exchange 'xs_guest'.

A subscriber can download the file http://dd.weather.gc.ca/bulletins/alphanumeric/SACN32_CWAO_123456 using http
without authentication on dd.weather.gc.ca.


ARGUMENTS AND OPTIONS
=====================

Please refer to the `sr_subscribe(1) <sr_subscribe.1.rst>`_ manual page for a detailed description of 
common settings, and methods of specifying them.

[-c|--config <configfile>]
--------------------------

  A list of settings in a configuration file 



[-p|--path path1 path2 ... pathN]
---------------------------------

  **sr_post** evaluates the filesystem paths from the **path** option 
  and possibly the **base_dir** if the option is used.

  If a path defines a file, this file is announced.

  If a path defines a directory, then all files in that directory are
  announced... 

[-pb|--post_broker <broker>]
----------------------------

  the broker to which the post is sent.

[-pbd|--post_base_dir <path>]
-----------------------------

  The *base_dir* option supplies the directory path that,
  when combined (or found) in the given *path*, 
  gives the local absolute path to the data file to be posted.
  The document root part of the local path will be removed from the posted announcement.
  For sftp URLs: it can be appropriate to specify a path relative to a user account.
  Example of that usage would be:  -dr ~user  -post_base_url sftp:user@host  
  For file URLs: base_dir is usually not appropriate.  To post an absolute path, 
  omit the -dr setting, and just specify the complete path as an argument.

[-px|--post_exchange <exchange>]
--------------------------------

  Sr_post publishes to an exchange named *xs_*"broker_username" by default.
  Use the *post_exchange* option to override that default.

[-h|-help|--help]
-----------------

  Display program options.

[--blocksize <value>]
---------------------

  This option controls the partitioning strategy used to post files.
  The value should be one of::

     0 - autocompute an appropriate partitioning strategy (default)
     1 - always send entire files in a single part.
     <blocksize> - used a fixed partition size (example size: 1M )

  Files can be announced as multiple parts.  Each part has a separate checksum.
  The parts and their checksums are stored in the cache. Partitions can traverse
  the network separately, and in parallel.  When files change, transfers are
  optimized by only sending parts which have changed.  
  
  The value of the *blocksize*  is an integer that may be followed by  letter designator *[B|K|M|G|T]* meaning:
  for Bytes, Kilobytes, Megabytes, Gigabytes, Terabytes respectively.  All these references are powers of 2.
  Files bigger than this value will get announced with *blocksize* sized parts.
  
  The autocomputation algorithm determines a blocksize that encourages a reasonable number of parts
  for files of various sizes.  As the file size varies, the automatic computation will give different
  results.  This will result in resending information which has not changed as partitions of a different 
  size will have different sums, and therefore be tagged as different.  
  
  By default, **sr_post** computes a reasonable blocksize that depends on the file size.
  The user can set a fixed *blocksize* if it is better for its products or if he wants to
  take advantage of the **cache** mechanism.  In cases where large files are being appended to, for example,
  it make sense to specify a fixed partition size so that the blocks in the cache will be the 

CHANGES.rst
===============
 Release Notes 
===============

lists all changes between versions.

**git repo**

**2.19.11b1**

  * bufix #280 added tls_rigour directive to help sr_poll deal with bad SSL certs.

**2.19.10b1**

  * bugfix #276 avoided repeat of #274 with self.length.
  * bugfix #273, #275, partflg regression in poll scripts
  * bugfix #270 from_cluster regression for older versions.
  * bugfix #268 new sr implementation broke python 3.4 compatibility.
  * documentation improvements.

**2.19.09b2**

  * documentation: usage string correction: replace parts by blocksize
  * documentation: other corrections.
  * windows package generation fixed (pending)
  * removed timeout on wait in process spawning plugins.

**2.19.09b1**

*    new: #191 adding log_reject to debug checksums and reject/accept configurations
*    new: #191 sr implementation replaced. old available as sr1. man page sr(8) added.
*    new: #227 some support added to for rabbitmq to work with mqtt.
*    new: tools to support migration of large Sundew configurations to Sarracenia.
*
* bugfix: 
*         #240,#254,#249,#252 logging and background launch output redirection.
*         #238 sr_watch crashes when encountering file it can't read
*         #212 fix for interpretation of username in baseurl (FIXME: requires review)
*         #211 fix for size approximation being wrong on polls that don't provide exact size.
*         #210 sr_audit race condition fixed by killing at the beginning of stop processing. 
*         #207 reassembly of partitioned files by sender.
*         #202 Stack overflow on sr_watch restart.
*         #201 fix for lookup in cache for partitioned files.
*         #190 when printing v03 messages, post_log was misleading (not updated)
*         #182 fixed support for file=NONE (Sundew compatibility)
*         #174 sr too slow fixed by replacing it.
*         also allowed removing some duplicate code. 
*         lost backward compatibility warnings around in on_post calls.
*         apparently, in Sundew keys: + is the same as .*  mapping added.
*
* testing improvements/fixes:
*         more design diagrams.
*         unit testing refactored.
*         #208 fixed: flow test is leaking...
*         #206 log rotation corrupts flow tests.
*         #184 some some flow test refactoring
*         #182 sr_watch fails on high volumes in flow test.
*   

**2.19.04b2**

* bugfix: #158 there was a bug with extended attributes in checksums (yes, another one.)
* new:    moved flow_test away from xpublic to allow continuous use
*         on an active pump without interfering with normal flows.
* info:   mostly refactoring for future work by Benoit.
*         Noureddine wants to run flow_test continuously somewhere
*         in addition to the travis-ci.com stuff.
*         there are some issues that crept in over the summer:
*         #182, #179, are opened to track progress on that.

**2.19.04b1**

* new:    adding WMO_mesh_post example.
* bugfix: regression all posts in b6 were v02 (v03 inadvertantly disabled.)
* bugfix: print improvements.
* bugfix: there was an issue with extended attribute in checksums.
* 

**2.19.03b6**

* new:    made code use instance variables instead of repeatedly parsing
*         elements of the message: topic_prefix->version, sum-> event.
*         more robust/easier to maintain.
* bugfix: updated WMO_mesh examples to use v03 and /var/www/html (as per feedback.)
* new:    add post_override_del option to post_override plugin.
*         allows deletion of headers on post.
* bugfix: Issue #175 documentation needs to use amqps for dd.weather, not amqp anymore.
* new:    wmo_mesh example now deletes a bunch of headers to shorten & simplify messages.
* new:    switching http port in flow test to 8001.  8000 is too popular.
* new:    switching flow_test to use hpfx.collab.science.gc.ca instead of dd.weather.gc.ca
* bugfix: Issue #168 have httpd unit test retrieve pick a file <= 2kb.  Big files 
*         were causing hangs in the flow test... this is one fix, but not enough to close
*         the bug.
* new:    Issue #54 is finally closed.

c/sr_post.c
/* vim:set ft=c ts=2 sw=2 sts=2 et cindent: */

/*
 * Usage info after license block.
 *
 * This code is by Peter Silva copyright (c) 2017 part of MetPX.
 * copyright is to the Government of Canada. code is GPLv2
 *
 * based on a amqp_sendstring from rabbitmq-c package
 * the original license is below:
 */

/* 
  Minimal c implementation to allow posting of sr_post(7) messages.
  It has a lot of limitations, and no error checking for now.

  how to use:

  In a shell, to use an sr_config(7) style configuration file:
  set the SR_POST_CONFIG environment variable to the name of the
  file to use.

 
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <strings.h>

#include <sys/types.h>
#include <sys/stat.h>

#include <dirent.h>

#include <unistd.h>
#include <fcntl.h>
#include <linux/limits.h>

#include <openssl/md5.h>
#include <openssl/sha.h>

#include <stdint.h>
#include <amqp_tcp_socket.h>
#include <amqp_ssl_socket.h>
#include <amqp.h>
#include <amqp_framing.h>

#include "sr_context.h"

// needed for sr_post_message.
#include "sr_consume.h"

#include "sr_version.h"

/*
 Statically assign the maximum number of headers that can be included in a message.
 just picked a number.  I remember picking a larger one before, and it bombed, don't know why.

 */
#define HDRMAX (255)

amqp_table_entry_t headers[HDRMAX];

int hdrcnt = 0 ;

void header_reset() {
    hdrcnt--;
    for(; (hdrcnt>=0) ; hdrcnt--)
    {
        headers[hdrcnt].key = amqp_cstring_bytes("");
        headers[hdrcnt].value.kind = AMQP_FIELD_KIND_VOID;
        headers[hdrcnt].value.value.bytes = amqp_cstring_bytes("");
    }
    hdrcnt=0;
}

void amqp_header_add( char *tag, const char * value ) {

  char value2[AMQP_MAX_SS];

  if ( hdrcnt >= HDRMAX ) 
  {
     log_msg( LOG_ERROR, "too many headers! ignoring %s=%s\n", tag, value );
     return;
  }
  headers[hdrcnt].key = amqp_cstring_bytes(tag);
  headers[hdrcnt].value.kind = AMQP_FIELD_KIND_UTF8;

  if ( strlen(value) > AMQP_MAX_SS ) 
  {
     strncpy( value2, value, AMQP_MAX_SS );
     value2[AMQP_MAX_SS] = '\0';
     log_msg( LOG_WARNING, "header %s too long (%d bytes), truncating to: %s\n", tag, strlen(value), value2 );
     headers[hdrcnt].value.value.bytes = amqp_cstring_bytes(value2);
  } else {
     headers[hdrcnt].value.value.bytes = amqp_cstring_bytes(value);
  }
  hdrcnt++;
  //log_msg( LOG_DEBUG, "Adding header: %s=%s hdrcnt=%d\n", tag, value, hdrcnt );
}

void set_url( char* m, char* spec ) 
  /* Pick a URL from the spec (round-robin) copy it to the given buffer
   */
{
  static const char* cu_url = NULL;
  char *sp;

  if ( strchr(spec,',') ) {
     //log_msg( LOG_DEBUG, "1 picking url, set=%s, cu=%s\n", spec, cu_url );
     if (cu_url) {
         cu_url = strchr(cu_url,','); // if there is a previous one, pick the next one.
         //log_msg( LOG_DEBUG, "2 picking url, set=%s, cu=%s\n", spec, cu_url );
     }
     if (cu_url) {
         cu_url++;                    // skip to after the comma.
         //log_msg( LOG_DEBUG, "3 picking url, set=%s, cu=%s\n", spec, cu_url );
     } else {
         cu_url = spec ;                // start from the beginning.
         //log_msg( LOG_DEBUG, "4 picking url, set=%s, cu=%s\n", spec, cu_url );
     }
     sp=strchr(cu_url,',');
     if (sp) strncpy( m, cu_url, sp-cu_url );
     else strcpy( m, cu_url );
  } else  {
     strcpy( m, spec );
  }
}

unsigned long int set_blocksize( long int bssetting, size_t fsz )
{
      unsigned long int tfactor =  (50*1024*1024) ;

      switch( bssetting )
      {
        case 0: // autocompute 
             if ( fsz > 100*tfactor ) return( 10*tfactor );
             else if ( fsz > 10*tfactor ) return( (unsigned long int)( (fsz+9)/10) );
             else if ( fsz > tfactor ) return( (unsigned long int)( (fsz+2)/3) ) ;
             else return(fsz);
             break;

        case 1: // send file as one piece.
             return(fsz);
             break;

       default: // partstr=i
             return(bssetting);
             break;
      }

}

void sr_post_message( struct sr_context *sr_c, struct sr_message_t *m )
{
    char fn[PATH_MAXNUL];
    char message_body[1024];
    char smallbuf[256];
    char thisexchange[256];
    char *c,*d;
    amqp_table_t table;
    amqp_basic_properties_t props;
    amqp_tx_commit_ok_t *commit_status;
    amqp_rpc_reply_t reply;
    signed int status;
    struct sr_header_t *uh;
    time_t to_sleep = 1;

    // MG white space in filename
    strcpy(fn, m->path);
    c = strchr(m->path,' ');
    if ( c != NULL )
    {
    c = fn;
    d = m->path;
    //while ( *d ) { if ( *d == ' ' ) { *c++='%'; *c++='2'; *c++='0'; } else *c++ = *d; d++; }

    while ( *d )
          {
          if ( *d == ' ' ) { *c++='%'; *c++='2'; *c++='0'; }
          else
             {
             if ( *d == '#' ) { *c++='%'; *c++='2'; *c++='3'; }
             else *c++ = *d;
             }
          d++;
          }
    *c='\0';
    }

    //  resume posting
    while(1) 
    {
        strcpy( message_body, m->datestamp);
        strcat( message_body, " " );
        strcat( message_body, m->url );
        strcat( message_body, " " );
        strcat( message_body, fn  );
        strcat( message_body, " \n" );
     
        header_reset();
    
        if ( m->from_cluster && m->from_cluster[0] )
            amqp_header_add( "from_cluster", m->from_cluster );
    
        if (( m->sum[0] != 'R' ) && ( m->sum[0] != 'L' ))
        {
           amqp_header_add( "parts", sr_message_partstr(m) );
    
           if ( m->atime && m->atime[0] )
               amqp_header_add( "atime", m->atime );
    
           if ( m->mode > 0 ) 
           {
               sprintf( smallbuf, "%04o", m->mode );
               amqp_header_add( "mode", smallbuf );
           }
    
           if ( m->mtime && m->mtime[0] )
               amqp_header_add( "mtime", m->mtime );
        }
    
        if ( m->sum[0] == 'L' )
        {
           amqp_header_add( "link", m->link );
        }
    
        amqp_header_add( "sum", m->sum );
    
        if ( m->to_clusters && m->to_clusters[0] )
            amqp_header_add( "to_clusters", m->to_clusters );
    
        for( uh=m->user_headers; uh ; uh=uh->next )
            amqp_header_add(uh->key, uh->value);
    
        table.num_entries = hdrcnt;
        table.entries=headers;
    
        props._flags = AMQP_BASIC_HEADERS_FLAG | AMQP_BASIC_CONTENT_TYPE_FLAG | AMQP_BASIC_DELIVERY_MODE_FLAG;
        props.content_type = amqp_cstring_bytes("text/plain");
        props.delivery_mode = 2; /* persistent delivery mode */
        props.headers = table;
    
        strcpy( thisexchange, sr_c->cfg->post_broker->exchange );
    
        if ( sr_c->cfg->post_broker->exchange_split > 0 ) 
        { 
            sprintf( strchr( thisexchange, '\0' ), "%02d", m->sum[ get_sumhashlen(m->sum[0])-1 ] % sr_c->cfg->post_broker->exchange_split ); 
        }
        status = amqp_basic_publish(sr_c->cfg->post_broker->conn, 1, amqp_cstring_bytes(thisexchange), 
                  amqp_cstring_bytes(m->routing_key), 0, 0, &props, amqp_cstring_bytes(message_body));
    
    
        if ( status < 0 ) 
        {
            log_msg( LOG_ERROR, "sr_%s: publish of message for  %s%s failed.\n", sr_c->cfg->progname, m->url, fn );
            goto restart;
        }
    
        commit_status = amqp_tx_commit( sr_c->cfg->post_broker->conn, 1 );
        if ( ! commit_status )
        {
           log_msg( LOG_ERROR, "broker failed to acknowledge publish event\n"  );
           reply = amqp_get_rpc_reply(sr_c->cfg->post_broker->conn);
           if (reply.reply_type != AMQP_RESPONSE_NORMAL ) {
               sr_amqp_reply_print(reply, "failed AMQP get_rpc_reply");
           }
           goto restart;
        }
        log_msg( LOG_INFO, "published: %s\n", sr_message_2log(m) );
        return;
    
restart:
        sr_context_close(sr_c);
        sleep(to_sleep);
        if (to_sleep < 60) to_sleep<<=1;
        log_msg( LOG_WARNING, "publish failed. Slept: %d seconds. Retrying...\n", to_sleep );
        sr_context_connect(sr_c);

    }
}

int sr_file2message_start(struct sr_context *sr_c, const char *pathspec, struct stat *sb, struct sr_message_t *m ) 
/*
  reading a file, initialize the message that corresponds to it. Return the number of messages to post entire file.
 */
{
  char  *drfound;
  char  fn[PATH_MAXNUL];
  char *c, *d;
  int lasti;
  int   linklen;
  char *linkp;
  char  linkstr[PATH_MAXNUL];
  char  tmprk[PATH_MAXNUL+100];
   
    if (*pathspec != '/' ) // need absolute path.
    { 
        getcwd( linkstr, PATH_MAX);
        strcat( linkstr, "/" );
        strcat( linkstr, pathspec);

        /* realpath stuff when it exists  sb */
        if ( sb && sr_c->cfg->realpath ) 
        {
            log_msg( LOG_DEBUG, "applying realpath to relpath %s\n", pathspec );
            if ( !realpath( linkstr, fn ) ) {
                 strcpy( fn, linkstr );
            }
        } else
            strcpy( fn, linkstr );
        linkstr[0]='\0';

    } else {
        /* realpath stuff when it exists  sb */
        if ( sb && sr_c->cfg->realpath ) 
        {
            log_msg( LOG_DEBUG, "applying realpath to abspath %s\n", pathspec );
            realpath( pathspec, fn );
        } else
            strcpy( fn, pathspec );
    }

  if ( (sr_c->cfg!=NULL) && sr_c->cfg->debug )
  {
     log_msg( LOG_DEBUG, "sr_%s file2message start with: %s sb=%p islnk=%d, isdir=%d, isreg=%d\n", 
         sr_c->cfg->progname, fn, sb, sb?S_ISLNK(sb->st_mode):0, sb?S_ISDIR(sb->st_mode):0,sb?S_ISREG(sb->st_mode):0 );
  }
  if ( sb && S_ISDIR(sb->st_mode) ) return(0); // cannot post directories.

  /* copy filename to path, but inserting %20 for every space
   */
  c = m->path;
  d = fn;
  //while ( *d ) { if ( *d == ' ' ) { *c++='%'; *c++='2'; *c++='0'; } else *c++ = *d; d++; }
  while ( *d )
        {
        if ( *d == ' ' ) { *c++='%'; *c++='2'; *c++='0'; }
        else
           {
           if ( *d == '#' ) { *c++='%'; *c++='2'; *c++='3'; }
           else *c++ = *d;
           }
        d++;
        }
  *c='\0';
  //strcpy( m->path, fn );

  if (sr_c->cfg->post_base_dir) 
  {
      drfound = strstr(fn, sr_c->cfg->post_base_dir ); 
   
      if (drfound) 
      {
          drfound += strlen(sr_c->cfg->post_base_dir) ; 
          strcpy( m->path, drfound );
      } 
  } 

  // use tmprk variable to fix  255 AMQP_SS_LEN limit
  strcpy( tmprk, sr_c->cfg->topic_prefix );
  strcat( tmprk, "." );
  strcat( tmprk, m->path + (*(m->path)=='/') );

  if ( strlen(tmprk) > 255 ) tmprk[255] = '\0';

  strcpy( m->routing_key, tmprk);


  lasti=0;
  for( int i=strlen(sr_c->cfg->topic_prefix) ; i< strlen(m->routing_key) ; i++ )
  {
      if ( m->routing_key[i] == '/' ) 
      {
           if ( lasti > 0 ) 
           {
              m->routing_key[lasti]='.';
           }
           lasti=i;
      }
  }
  m->routing_key[lasti]='\0';

  strcpy( m->datestamp, sr_time2str(NULL));
  strcpy( m->to_clusters, sr_c->cfg->to );

  m->parts_blkcount=1;
  m->parts_rem=0;
  m->parts_num=0;

  m->user_headers=sr_c->cfg->user_headers;

  m->sum[0]= sr_c->cfg->sumalgo;


  if ( !sb ) 
  {
      if ( ! ((sr_c->cfg->events)&SR_DELETE) ) return(0); // not posting deletes...
      m->sum[0]='R';
  } else if ( S_ISLNK(sb->st_mode) ) 
  {
      if ( ! ((sr_c->cfg->events)&SR_LINK) ) return(0); // not posting links...

      strcpy( m->atime, sr_time2str(&(sb->st_atim)));
      strcpy( m->mtime, sr_time2str(&(sb->st_mtim)));
      m->mode = sb->st_mode & 07777 ;

      m->sum[0]='L';
      linkstr[0]='\0';
      linklen = readlink( fn, linkstr, PATH_MAX );
      linkstr[linklen]='\0';
      if ( sr_c->cfg->realpath ) 
      {
          linkp = realpath( linkstr, m->link );
          if (!linkp) 
          {
               log_msg( LOG_ERROR, "sr_%s unable to obtain realpath for %s\n", sr_c->cfg->progname, fn );
               return(0);
          }
      } else {
         strcpy( m->link, linkstr ); 
      }

  } else if (S_ISREG(sb->st_mode)) 
  {   /* regular files, add mode and determine block parameters */

      if ( ! ((sr_c->cfg->events)&(SR_CREATE|SR_MODIFY)) ) return(0);

      if ( access( fn, R_OK ) ) return(0); // will not be able to checksum if we cannot read.

      strcpy( m->atime, sr_time2str(&(sb->st_atim)));
      strcpy( m->mtime, sr_time2str(&(sb->st_mtim)));
      m->mode = sb->st_mode & 07777 ;

      m->parts_blksz  = set_blocksize( sr_c->cfg->blocksize, sb->st_size );
      m->parts_s = (m->parts_blksz < sb->st_size )? 'i':'1' ;

      if ( m->parts_blksz == 0 ) {
          m->parts_rem = 0;
      } else {
          m->parts_rem = sb->st_size%(m->parts_blksz) ;
          m->parts_blkcount = ( sb->st_size / m->parts_blksz ) + ( m->parts_rem?1:0 );
      }

  } 
  return(m->parts_blkcount);
}

struct sr_message_t *sr_file2message_seq(const char *pathspec, int seq, struct sr_message_t *m ) 
/*
  Given a message from a "started" file, the prototype message, and a sequence number ( sequence is number of blocks of partsze )
  return the adjusted prototype message.  (requires reading part of the file to checksum it.)
 */
{
      m->parts_num = seq;

      strcpy( m->sum, 
              set_sumstr( m->sum[0], pathspec, NULL, m->link, m->parts_blksz, m->parts_blkcount, m->parts_rem, m->parts_num ) 
            ); 

      if ( !(m->sum) ) 
      {
         log_msg( LOG_ERROR, "file2message_seq unable to generate %c checksum for: %s\n", m->parts_s, pathspec );
         return(NULL);
      }
  return(m);
}

void sr_post(struct sr_context *sr_c, const char *pathspec, struct stat *sb ) 
{
  static struct sr_message_t m;
  int numblks;
  int status;

  strcpy( m.to_clusters, sr_c->cfg->to );
  strcpy( m.from_cluster, sr_c->cfg->post_broker->hostname );
  strcpy( m.source,  sr_c->cfg->source );
  set_url( m.url, sr_c->cfg->post_base_url );
  m.user_headers = sr_c->cfg->user_headers;

  // report...
  // FIXME: duration, consumingurl, consuminguser, statuscode?
  numblks = sr_file2message_start( sr_c, pathspec, sb, &m );

  for( int blk=0; (blk < numblks); blk++ )
  {
      if ( sr_file2message_seq(pathspec, blk, &m ) ) 
      {
           if ( sr_c->cfg->cache > 0 ) 
           { 
               status = sr_cache_check( sr_c->cfg->cachep, m.sum[0], (unsigned char*)(m.sum), m.path, sr_message_partstr(&m) ); 
               log_msg( LOG_DEBUG, "sr_post cache_check: %s\n", status?"not found":"already there, no post" );
               if (!status) continue; // cache hit.
           }
          sr_post_message( sr_c, &m );
      }
  }

}

void sr_post_rename(struct sr_context *sr_c, const char *oldname, const char *newname);

void sr_post_rename_dir(struct sr_context *sr_c, const char *oldname, const char *newname)
{
  DIR *dir;
  struct dirent *e;
  char oldpath[PATH_MAX+1];
  int oldlen;
  char newpath[PATH_MAX+1];
  int newlen;

  log_msg( LOG_DEBUG, "sr_%s %s starting rename_dir: %s %s \n", sr_c->cfg->progname, __sarra_version__, oldname, newname );
  dir = opendir( newname );
  if (!dir) return;
  
  // build oldname and new name as sub of current ones.
  strcpy(oldpath, oldname );
  strcat(oldpath, "/" );
  oldlen=strlen(oldpath);

  strcpy(newpath, newname );
  strcat(newpath, "/" );
  newlen = strlen(newpath);
 
  while( (e = readdir(dir)) ) 
  {
      if ( !strcmp(e->d_name,".") || !strcmp(e->d_name,"..") )
          continue;
      strcat( oldpath, e->d_name );
      strcat( newpath, e->d_name );
      sr_post_rename( sr_c, oldpath, newpath );

      oldpath[oldlen]='\0';
      newpath[newlen]='\0';
  }

  closedir(dir);
}

void sr_post_rename(struct sr_context *sr_c, const char *o, const char *n)
/*
   assume actual rename is completed, so newname exists.
 */
{
  struct stat sb;
  struct sr_header_t first_user_header;
  struct sr_mask_t *mask;
  char *s;
  char oldname[PATH_MAX];
  char oldreal[PATH_MAX];
  char newname[PATH_MAX];
  char newreal[PATH_MAX];
  char tmpname[PATH_MAX];

  strcpy( oldname, o );
  strcpy( newname, n );

  if (sr_c->cfg->realpath || sr_c->cfg->realpath_filter) {
     strcpy( tmpname, o );
     s=rindex( tmpname, '/' );
     if (s) 
     {
        *s='\0';
        s++;
     }
     realpath( tmpname, oldreal ); // apply to directory, not final path entry.
     if (s)
     {
        strcat( oldreal, "/" );
        strcat( oldreal, s );
     }
     log_msg( LOG_DEBUG, "applying realpath to old: %s -> %s\n", o, oldreal );

     realpath( n, newreal);
     log_msg( LOG_DEBUG, "applying realpath to new: %s -> %s\n", n, newreal );
  } 

  if (sr_c->cfg->realpath ) {
     strcpy( oldname, oldreal );
     strcpy( newname, newreal );
  }
      
  log_msg( LOG_DEBUG, "sr_%s %s starting rename: %s %s \n", sr_c->cfg->progname, __sarra_version__, oldname, newname );

  if ( lstat( newname, &sb ) ) 
  {
     log_msg( LOG_ERROR, "sr_%s rename: %s cannot stat.\n", sr_c->cfg->progname, newname );
     return;
  }
  if ( S_ISDIR(sb.st_mode) )
  {
      sr_post_rename_dir( sr_c, oldname, newname );
  }

  first_user_header.next = sr_c->cfg->user_headers;
  sr_c->cfg->user_headers =  &first_user_header ;

  first_user_header.key = strdup( "newname" );
  first_user_header.value = strdup( newname );

  if ( sr_c->cfg->realpath_filter) {
    mask = isMatchingPattern( sr_c->cfg, oldreal );
  } else {
    mask = isMatchingPattern( sr_c->cfg, oldname );
  }
  if ( (mask && !(mask->accepting)) || (!mask && !(sr_c->cfg->accept_unmatched)) ) 
      log_msg( LOG_DEBUG, "rejecting: %s\n", oldname );
  else {
       if ( !access(oldname, F_OK) && (S_ISREG(sb.st_mode) || S_ISLNK(sb.st_mode) ) ) {
            sr_post( sr_c,  oldname, &sb  );
       } else {
            sr_post( sr_c,  oldname, NULL );
       }
  }

  free(first_user_header.key);  
  free(first_user_header.value);  
  first_user_header.key = strdup( "oldname" );
  first_user_header.value = strdup( oldname );

  if ( sr_c->cfg->realpath_filter) {
     mask = isMatchingPattern( sr_c->cfg, newreal );
  } else {
    mask = isMatchingPattern( sr_c->cfg, newname );
  }
  if ( (mask && !(mask->accepting)) || (!mask && !(sr_c->cfg->accept_unmatched)) ) 
      log_msg( LOG_DEBUG, "rejecting: %s\n", newname );
  else 
      sr_post( sr_c,  newname, &sb );

  free(first_user_header.key);  
  sr_c->cfg->user_headers = first_user_header.next ;
  

}

int sr_post_cleanup( struct sr_context *sr_c )
{
    char exchange[256];
    amqp_rpc_reply_t reply;

    if ( sr_c->cfg->post_broker->exchange_split ) 
    {
        for ( int i = 0; i < sr_c->cfg->post_broker->exchange_split ; i++ )
        {
            sprintf( exchange, "%s%02d", sr_c->cfg->post_broker->exchange, i );
            log_msg( LOG_INFO, "deleting exchange %s%02d\n", sr_broker_uri( sr_c->cfg->post_broker ), i  );
            amqp_exchange_delete( sr_c->cfg->post_broker->conn, 1, amqp_cstring_bytes(exchange), 0 );
            reply = amqp_get_rpc_reply(sr_c->cfg->post_broker->conn);
            if (reply.reply_type != AMQP_RESPONSE_NORMAL ) 
            {
                sr_amqp_reply_print(reply, "failed AMQP get_rpc_reply exchange delete");
            }
        }
    } else {
        log_msg( LOG_INFO, "deleting exchange %s\n", sr_broker_uri( sr_c->cfg->post_broker )  );
        amqp_exchange_delete( sr_c->cfg->post_broker->conn, 1, amqp_cstring_bytes(sr_c->cfg->post_broker->exchange), 0 );
        reply = amqp_get_rpc_reply(sr_c->cfg->post_broker->conn);
        if (reply.reply_type != AMQP_RESPONSE_NORMAL ) 
        {
            sr_amqp_reply_print(reply, "failed AMQP get_rpc_reply exchange delete");
        }
    }
    return(1);
}

int sr_post_init( struct sr_context *sr_c )
{
    char exchange[256];
    amqp_rpc_reply_t reply;


    if ( sr_c->cfg->post_broker->exchange_split ) 
    {
        for ( int i = 0; i < sr_c->cfg->post_broker->exchange_split ; i++ )
        {
            log_msg( LOG_INFO, "declaring exchange %s%02d\n", sr_broker_uri( sr_c->cfg->post_broker ), i  );
            sprintf( exchange, "%s%02d", sr_c->cfg->post_broker->exchange, i );
            amqp_exchange_declare( sr_c->cfg->post_broker->conn, 1, amqp_cstring_bytes(exchange),
                 amqp_cstring_bytes("topic"), 0, sr_c->cfg->durable, 0, 0, amqp_empty_table );
            reply = amqp_get_rpc_reply(sr_c->cfg->post_broker->conn);
            if (reply.reply_type != AMQP_RESPONSE_NORMAL ) 
            {
                sr_amqp_reply_print(reply, "failed AMQP get_rpc_reply exchange declare");
            }
        }
    } else  {
        log_msg( LOG_INFO, "declaring exchange %s\n", sr_broker_uri( sr_c->cfg->post_broker )  );
        amqp_exchange_declare( sr_c->cfg->post_broker->conn, 1, amqp_cstring_bytes(sr_c->cfg->post_broker->exchange),
             amqp_cstring_bytes("topic"), 0, sr_c->cfg->durable, 0, 0, amqp_empty_table );
        reply = amqp_get_rpc_reply(sr_c->cfg->post_broker->conn);
        if (reply.reply_type != AMQP_RESPONSE_NORMAL ) 
        {
            sr_amqp_reply_print(reply, "failed AMQP get_rpc_reply exchange declare");
        }
    } 

    return(1);
}

